fastestimator/trace/adapt/reduce_lr_on_plateau.py

```diff
@@ -0,0 +1,82 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+import inspect
+from typing import Optional, Union
+
+import numpy as np
+import tensorflow as tf
+import torch
+
+from fastestimator.backend import get_lr, set_lr
+from fastestimator.summary import System
+from fastestimator.trace.trace import Trace
+from fastestimator.util.data import Data
+
+
+class ReduceLROnPlateau(Trace):
+    """Reduce learning rate based on evaluation results
+
+    Args:
+        model: model instance
+        monitor: The monitor metric name, if None, model's validation loss will be used as metric.
+        patience: Number of epochs to wait before reducing LR again.
+        factor: Reduce factor of learning rate. Defaults to 0.1.
+        reduce_mode: higher is better (max) or lower is better (min). Defaults to min
+        min_lr: minimum learning rate, defaults to 1e-6.
+    """
+    system: System
+
+    def __init__(self,
+                 model: Union[tf.keras.Model, torch.nn.Module],
+                 metric: Optional[str] = None,
+                 patience: int = 10,
+                 factor: float = 0.1,
+                 reduce_mode: str = "min",
+                 min_lr: float = 1e-6):
+        if not metric:
+            assert hasattr(model, "loss_name"), "cannot infer model loss name, please put the model to UpdateOp first"
+            assert len(model.loss_name) == 1, "the model has more than one losses, please provide the metric explicitly"
+            metric = next(iter(model.loss_name))
+        super().__init__(mode="eval", inputs=metric)
+        self.model = model
+        self.patience = patience
+        self.factor = factor
+        self.reduce_mode = reduce_mode
+        self.min_lr = min_lr
+        self.wait = 0
+        if self.reduce_mode == "min":
+            self.best = np.Inf
+            self.monitor_op = np.less
+        elif self.reduce_mode == "max":
+            self.best = -np.Inf
+            self.monitor_op = np.greater
+        else:
+            raise ValueError("reduce_mode must be either 'min' or 'max'")
+
+    @property
+    def metric(self):
+        return self.inputs[0]
+
+    def on_epoch_end(self, data: Data):
+        if self.monitor_op(data[self.metric], self.best):
+            self.best = data[self.metric]
+            self.wait = 0
+        else:
+            self.wait += 1
+            if self.wait >= self.patience:
+                new_lr = max(self.min_lr, np.float32(self.factor * get_lr(self.model)))
+                set_lr(self.model, new_lr)
+                self.wait = 0
+                print("FastEstimator-ReduceLROnPlateau: learning rate reduced to {}".format(new_lr))
```

Comment:
I think we want to put this into the logs since they're now integrated with out experiment summary / auditing stuff

Reply:
done